### ISSUE TITLE
OpenTelemetry not working on the sink

### ISSUE THREAD

**Issue by rh-hemartin** [kubearchive/kubearchive#158]
When I deploy KubeArchive with values `apiServer.observability: true` and `sink.observability: true` I expect to see opentelemetry traces in the logs. That happens with the api, but not with the sink. Everything looks good and there is no error from the sink, however it does not print traces on stderr. I checked the deployments and they have the correct environment variables.

**Comment by rh-hemartin** [comment]
Knowing this we can do testing and understand what is happening.

**Comment by rh-hemartin** [comment]
I can't follow very well the logic of opentelemetry through the different packages. Its difficult to figure out what is happening. One thing I found out is that the middleware is just executed once, because I introduced a `logger.Println()` on it and only shows once no matter how many messages get received.

I think its worth to spend some time to switch to use Gin and OtelGin so we have both sink and API aligned on that. What do you think?

**Comment by rh-hemartin** [comment]
I think the problem is on the HTTP client creation, I'm investigating deeper:

```
httpClient, err := cloudevents.NewHTTP(
		cloudevents.WithRoundTripper(otelhttp.NewTransport(http.DefaultTransport)),
		cloudevents.WithMiddleware(func(next http.Handler) http.Handler {
			return otelhttp.NewHandler(next, "receive")
		}),
	)
```

Looks like removing the `WithMiddleware` option allows the `WithObservabilityService` work properly.

**Comment by rh-hemartin** [comment]
Deleting the WithMiddleware makes the WithObservabilityService telemetry work, as the spans that get sent are not named "receive".

**Comment by rh-hemartin** [comment]
I think its the traceparent header that gets sent from the ApiServerSource.

I introduced a logging statement for the headers of the request, I compared the headers from both curl and ApiServerSource calls. The ApiServerSource sends a traceparent headers which is part of the [specification for traces](https://www.w3.org/TR/trace-context/#traceparent-header).

I sent the curl command with that traceparent and no trace was printed. Then I switched the last bit of the traceparent  header (from 00 to 01)  and I got the traces. This last bit is the `sampled` flag which indicates if the parent recorded that trace. The `sampled` flag prevents broken traces where some parts where not recorded, so if the flag is `00` (not sampled) then the trace library should not record it. When the sampled flag is `01` (sampled) the library records (outputs in our case) the sample.

We should be able to confirm this by enabling tracing on the ApiServerSource. That way the header should have the sampled flag set to 01 and our tracing should be printed.

```
curl -v  \
  -H "Ce-Id: e7d95c20-6eb4-4614-946d-27b0ce41c7ff" \
  -H "Ce-Source: /apis/namespaces/dimitar/clone-build-n4qhgl" \
  -H "Ce-Subject: clone-build-n4qhgl" \
  -H "Ce-Specversion: 1.0" \
  -H "Ce-Type: dev.tekton.event.pipelinerun.started.v1" \
  -H "Content-Type: application/json" \
  -H "Ce-apiversion: tekton.dev/v1beta1" \
  -H "ce-kind: PipelineRun" \
  -H "ce-name: clone-build-n4qhgl" \
  -H "ce-namespace: dimitar" \
  -H "traceparent: 00-e447492abfc509239ffa680bc7eadcf1-3dc9598087914194-00" \
  -d @event.json http://localhost:8080
```

**Comment by rh-hemartin** [comment]
TLDR: our traces are working correctly. They are not being printed because the ApiServerSource is not recording its traces. If we enable tracing on the ApiServerSource our traces should be printed. This mechanism guarantees there are no broken traces (where some spans are recorded and some are not).

**Comment by rh-hemartin** [comment]
I created #199 to enable traces on Knative Eventing by default, so we get the traces.

**Comment by rh-hemartin** [comment]
Confirmed that our traces work when Knative Eventing traces are enabled.

**Comment by skoved** [comment]
That's interesting that ApiServerSource sets a traceparent header even if trace has not been enable in knative-eventing

**Comment by rh-hemartin** [comment]
> That's interesting that ApiServerSource sets a traceparent header even if trace has not been enable in knative-eventing

I didn't think of this... I will reach the knative eventing folks and see if this sounds good to them.
